linalg/ginkgo.hpp

```diff
@@ -31,8 +31,203 @@
 
 namespace mfem
 {
-namespace GinkgoWrappers
+namespace Ginkgo
 {
+/**
+* Helper class for a case where a wrapped MFEM Vector
+* should be owned by Ginkgo, and deleted when the wrapper
+* object goes out of scope.
+*/
+template <typename T>
+class gko_mfem_destroy
+{
+public:
+   using pointer = T *;
+
+   // Destroys an MFEM object.  Requires object to have a Destroy() method.
+   void operator()(pointer ptr) const noexcept { ptr->Destroy(); }
+};
+
+/**
+* This class wraps an MFEM vector object for Ginkgo's use.  It
+* is allows Ginkgo and MFEM to operate directly on the same
+* data, and is necessary to use MFEM Operators with Ginkgo
+* solvers.
+*
+* @ingroup Ginkgo
+*/
+
+class VectorWrapper : public gko::matrix::Dense<double>
+{
+public:
+   VectorWrapper(std::shared_ptr<const gko::Executor> exec,
+                 gko::size_type size, Vector *mfem_vec,
+                 bool ownership = false)
+      : gko::matrix::Dense<double>(
+           exec, gko::dim<2> {size, 1},
+   gko::Array<double>::view(exec, size,
+                            mfem_vec->ReadWrite(
+                               exec != exec->get_master() ? true : false)),
+   1)
+   {
+      // This controls whether or not we want Ginkgo to own its MFEM Vector.
+      // Normally, when we are wrapping an MFEM Vector created outside
+      // Ginkgo, we do not want ownership to be true. However, Ginkgo
+      // creates its own temporary vectors as part of its solvers, and
+      // these will be owned (and deleted) by Ginkgo.
+      if (ownership)
+      {
+         using deleter = gko_mfem_destroy<Vector>;
+         wrapped_vec = std::unique_ptr<Vector,
+         std::function<void(Vector *)>>(
+            mfem_vec, deleter{});
+      }
+      else
+      {
+         using deleter = gko::null_deleter<Vector>;
+         wrapped_vec = std::unique_ptr<Vector,
+         std::function<void(Vector *)>>(
+            mfem_vec, deleter{});
+      }
+   }
+   static std::unique_ptr<VectorWrapper> create(
+      std::shared_ptr<const gko::Executor> exec, gko::size_type size,
+      Vector *mfem_vec, bool ownership = false)
+   {
+      return std::unique_ptr<VectorWrapper>(
+                new VectorWrapper(exec, size, mfem_vec, ownership));
+   }
+   // Return reference to MFEM Vector object
+   Vector &get_mfem_vec_ref() const { return *(this->wrapped_vec.get()); }
+   // Return const reference to MFEM Vector object
+   const Vector &get_mfem_vec_const_ref() const
+   {
+      return const_cast<const Vector &>(*(this->wrapped_vec.get()));
+   }
```

Comment:
You could also potentially rename `get_mfem_vec_const_ref` to `get_mfem_vec_ref` and use overloading.

Reply:
Updated in 43f0cf1cc99410599b5f04440c16adc44faa7c7c

Reply:
You didn't like the idea to rename `get_mfem_vec_const_ref` to `get_mfem_vec_ref`?

Reply:
I do have a preference for leaving it the way it is, for two reasons:

1) I mainly view the `VectorWrapper` and `OperatorWrapper` classes as specialized extensions of Ginkgo itself.  `VectorWrapper`, for example, inherits from `gko::matrix::Dense`.  These two classes use Ginkgo-style error/exception throwing (see, e.g., `create_submatrix_impl`) and implement several virtual functions overriding their base classes in Ginkgo.   Basically, I was thinking that though the `*Wrapper` classes are available, and enterprising MFEM users could create objects from them directly if a need arose, they are mostly support classes for the internal benefit of the solver and preconditioner classes -- which is why I left them more "Ginkgo-like" than "MFEM-like". Anyway, Ginkgo uses this style, with`get_const_*` for the `const` versions of `get` functions, rather than overloading.  

2) I kind of like that it makes it very clear whether you are getting a `const` reference from the wrapper or not, when these functions are used (e.g. to get MFEM `Vectors` to pass to `Mult`).

All of that being said, I'm not so strongly tied to it that I would refuse to change it, if MFEM really wants it changed.  That was just my thought process behind it.